File: sdk/java/src/main/java/com/anyvali/schemas/RecordSchema.java
package com.anyvali.schemas;

import com.anyvali.IssueCodes;
import com.anyvali.Schema;
import com.anyvali.ValidationContext;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Schema for record/map with string keys and uniform value schema.
 */
public class RecordSchema extends Schema {
    private final Schema valueSchema;

    public RecordSchema(Schema valueSchema) {
        this.valueSchema = valueSchema;
    }

    private RecordSchema(RecordSchema other) {
        this.coercion = other.coercion;
        this.defaultValue = other.defaultValue;
        this.hasDefault = other.hasDefault;
        this.valueSchema = other.valueSchema;
    }

    @Override
    @SuppressWarnings("unchecked")
    protected Object validate(Object input, ValidationContext ctx) {
        if (!(input instanceof Map<?, ?> rawMap)) {
            String received = anyvaliTypeName(input);
            ctx.addIssue(IssueCodes.INVALID_TYPE,
                    "Expected object, received " + received, "object", received);
            return null;
        }

        var result = new LinkedHashMap<String, Object>();
        for (var entry : rawMap.entrySet()) {
            Object key = entry.getKey();
            if (!(key instanceof String sKey)) {
                ctx.addIssue(IssueCodes.INVALID_TYPE,
                        "Record keys must be strings, received " + key.getClass().getSimpleName(),
                        "string", key.getClass().getSimpleName());
                continue;
            }
            var childCtx = ctx.child(sKey);
            Object parsed = valueSchema.runPipeline(entry.getValue(), childCtx);
            result.put(sKey, parsed);
        }
        return result;
    }

    @Override
    protected Map<String, Object> toNode() {
        var node = new LinkedHashMap<String, Object>();
        node.put("kind", "record");
        node.put("values", valueSchema.toNode());
        return addCommonNodeFields(node);
    }

    @Override
    protected Schema copy() {
        return new RecordSchema(this);
    }

    public Schema getValueSchema() {
        return valueSchema;
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/schemas/RecordSchema.kt
package com.anyvali.schemas

import com.anyvali.*
import kotlinx.serialization.json.*

data class RecordSchema(
    val values: Schema<*>
) : Schema<Map<String, Any?>>() {
    override val kind: String = "record"

    override fun safeParseWithContext(input: Any?, ctx: ValidationContext): ParseResult<Map<String, Any?>> {
        if (input !is Map<*, *>) {
            return ParseResult.Failure(
                listOf(
                    ValidationIssue(
                        code = IssueCodes.INVALID_TYPE,
                        path = ctx.path,
                        expected = "record",
                        received = getJsonTypeName(input)
                    )
                )
            )
        }

        @Suppress("UNCHECKED_CAST")
        val inputMap = input as Map<String, Any?>
        val issues = mutableListOf<ValidationIssue>()
        val result = mutableMapOf<String, Any?>()

        for ((key, value) in inputMap) {
            val childCtx = ctx.child(key)
            val childResult = values.safeParseWithContext(value, childCtx)
            when (childResult) {
                is ParseResult.Success -> result[key] = childResult.value
                is ParseResult.Failure -> issues.addAll(childResult.issues)
            }
        }

        return if (issues.isEmpty()) ParseResult.Success(result) else ParseResult.Failure(issues)
    }

    override fun validateValue(value: Any?, ctx: ValidationContext): List<ValidationIssue> {
        val result = safeParseWithContext(value, ctx)
        return if (result is ParseResult.Failure) result.issues else emptyList()
    }

    override fun exportNode(): JsonObject = buildJsonObject {
        put("kind", JsonPrimitive("record"))
        put("values", values.exportNode())
    }
}
